[CASSANDRA-20132][trunk] Add new table metric PurgeableTombstoneScannedHistogram and a tracing event for scanned purgeable tombstones

netudima opened the pull request:
Add new table metric PurgeableTombstoneScannedHistogram and a tracing event for scanned purgeable tombstones

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink for [CASSANDRA-20132](https://issues.apache.org/jira/browse/CASSANDRA-20132)

Commit:
Address some of review comments

Patch by Dmitry Konstantinov; reviewed by Štefan Miklošovič for CASSANDRA-19818

netudima committed:
Add new table metric PurgeableTombstoneScannedHistogram and a tracing event for scanned purgeable tombstones

Patch by Dmitry Konstantinov; reviewed by TBD for CASSANDRA-20132

Commit:
Address review comments: add the configuration description to cassandra.yaml, add an explicit type cast check

Patch by Dmitry Konstantinov; reviewed by Štefan Miklošovič for CASSANDRA-20132

Commit:
Address review comments: enable considerZeros for the new metric, add partition tombstone counting
Add purgeable_tobmstones_metric_granularity configuration option to give the ability to control an overhead of the added metric collection, use row mode as a default to reduce overheads
Make added unit tests more readable

Patch by Dmitry Konstantinov; reviewed by Chris Lohfink for CASSANDRA-20132

smiklosovic commented on src/java/org/apache/cassandra/db/ReadCommand.java:
```diff
@@ -894,6 +896,134 @@ protected LongPredicate getPurgeEvaluator()
+            @Override
+            public void onClose()
+            {
+                cfs.metric.purgeableTombstoneScannedHistogram.update(purgeableTombstones);
```
what if `purgeableTombstones` is 0? Then we update with 0? Does it make sense to update with 0? I guess it does as it is histogram ... What if we wrapped both this call and tracing in that `if` one line below?

netudima replied:
We have a histogram metric here, so we want to know the whole distribution, if we skip 0 values then it will be biased and metrics like 90% percentiles will not reflect the reality (for example, you have all values except one as 0 and one is lets say 200, so the histogram will say in this case that median for the number of purgeable tombstones is 200). Other histogram metrics like plain tombstones capture 0 values as well.

Tracing is not about statistics aggregation but capturing details for individual queries, so for tracing such skips helps to reduce overheads and output verbosity without misleading a user.

smiklosovic commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -544,6 +544,8 @@ public static class SSTableConfig
     public volatile int tombstone_warn_threshold = 1000;
     public volatile int tombstone_failure_threshold = 100000;
 
+    public TombstonesMetricGranularity purgeable_tobmstones_metric_granularity = TombstonesMetricGranularity.row;
```
@netudima why does not this have documented configuration property in `cassandra.yaml` and `cassandra_latest.yaml`? Is there some reason we are not exposing this property to a user to use so we have it just here, undocumented?

smiklosovic replied:
@netudima there is also a typo in `tobmstones`.

smiklosovic replied:
I would most probably also rename this property. As of now, if you wake me up in the middle of the night and ask under what name the property was added I could not tell :) But I could tell you that it has something common _with tombstones_.

There are two tombstones properties just above starting on `tombstone` so I would do the same here. Maybe call it like `tombstone_purge_metric_granularity`.

netudima replied:
> why does not this have documented configuration property in cassandra.yaml and cassandra_latest.yaml?

Sorry,I just missed it, there was no an intentional reason. I will add it. Maybe we should think about an automatic test to check such consistency (as a part of a separate ticket, of course)

> there is also a typo in tobmstones.

good catch, thank you!

> tombstone_purge_metric_granularity

it sounds a bit strange for me because purgeable was an adjective here and now it looks more like a verb, so people may think that the metric is about a purging activity itself .. But I do not see a problem to change the name to tombstone_purge_metric_granularity if you think it is better.

smiklosovic replied:
yeah please change it to that ... or, ah, I dont know :D try to gather some basic feedback e.g. on Slack.

smiklosovic commented on src/java/org/apache/cassandra/db/ReadCommand.java:
```diff
@@ -894,6 +896,134 @@ protected LongPredicate getPurgeEvaluator()
+                final long nowInSec = nowInSec();
+                boolean hasTombstones = false;
+
+                if (DatabaseDescriptor.getPurgeableTobmstonesMetricGranularity() == Config.TombstonesMetricGranularity.cell)
```
why are we doing this every time we apply to row? This might be a private property resolved in `PurgeableTombstonesMetricRecording`'s constructor and here it would be just a boolean.

Could you also wrap that whole `for` in braces of `if`?

netudima replied:
> Could you also wrap that whole for in braces of if?

yes, np

> This might be a private property resolved in PurgeableTombstonesMetricRecording's constructor and here it would be just a boolean.

If it is a performance optimization then the cost of this operation negligible, there is no any type conversion here (like we have in some places where a set value is in human format like 32 MiB and get value is needed in bytes), so here we actually speak about a difference between reference comparison vs boolean check.
If we speak about this level of micro-optimization - adding a boolean field will increase PurgeableTombstonesMetricRecording object size for a typical case (OpenJDK 8/11/17, 8-byte alignment: 12 bytes header + 4 bytes for existing purgeableTombstones + 1 byte for the new field = 17 bytes -- alignment --> 24 bytes; so, we will increase our object size from the current 16 bytes to 24 bytes).

Another way is to set the condition result to a static variable but it complicates unit testing.

smiklosovic replied:
yes, I am just talking about readability mostly. Up to you.

netudima replied:
in case of readability what's about a static private method, like: boolean isPurgeableCellTombstonesTrackingEnabled() ?

smiklosovic replied:
sure

smiklosovic commented on src/java/org/apache/cassandra/db/ReadCommand.java:
```diff
@@ -894,6 +896,134 @@ protected LongPredicate getPurgeEvaluator()
+                // for bound markers - just increment if it is purgeable
+                else
+                {
+                    countIfPurgeable(((RangeTombstoneBoundMarker) marker).deletionTime(), nowInSec);
```
Bold cast! How do you know it is always like that? I see that when `isBoundary()` returns false, that is just in case of `RangeTombstoneBoundMarker` but it feels weird to base that on that logic.

smiklosovic replied:
@netudima could we at least check that `marker` is instance of `RangeTombstoneBoundMarker` before casting? If it is not instance of `RangeTombstoneBoundMarker`, we should just skip this logic.

netudima replied:
Yes, I am planning to do this, just have not committed yet

netudima replied:
done

smiklosovic commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -1327,6 +1329,25 @@ public enum BatchlogEndpointStrategy
         }
     }
 
+    public enum TombstonesMetricGranularity
```
put a comment on this like "see CASSANDRA-20132" etc ...

smiklosovic commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -1327,6 +1329,25 @@ public enum BatchlogEndpointStrategy
+        row,
+        /**
+         * track partition/range/row/cell level tombstones,
+         * the most granular option, but it has some performance overheads
```
It would be nice to quantify what "some overheads" means.  How much slower will it be? So a user knows what he is buying into. Should be visible in cassandra.yaml as well. Maybe wording like "slower as in matter of single digit percents" or so ... You got it.

smiklosovic commented on src/java/org/apache/cassandra/db/ReadCommand.java:
```diff
@@ -466,6 +466,8 @@ public UnfilteredPartitionIterator executeLocally(ReadExecutionController execut
                 iterator = withQuerySizeTracking(iterator);
                 iterator = maybeSlowDownForTesting(iterator);
                 iterator = withQueryCancellation(iterator);
+                if (DatabaseDescriptor.getPurgeableTobmstonesMetricGranularity() != Config.TombstonesMetricGranularity.disabled)
```
for readability's sake, I would do `maybeRecordPurgeableTombstones` and in that method there would be this `if` and if not true then I would just return `iter`. The same way it is done for `maybeSlowDownForTest` two lines above.

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -1964,6 +1964,20 @@ transparent_data_encryption_options:
+#              but it has some performance overheads due to iteration over cells.
+#              For CPU-bound workload you may get about 5% of overhead for throughput.
+#              For IO-bound workload the overhead is almost negligible.
+# tombstone_read_purgeable_metric_granularity: row
```
We need to disable this by default (as well in latest yaml). Same with `Config.java`. You need to opt-in into this.

smiklosovic replied:
@netudima btw could you please add this metric to `doc/modules/cassandra/pages/managing/operating/metrics.adoc` as well?

Commit:
Disable the new metric by default, document the metric in metrics.adoc

Patch by Dmitry Konstantinov; reviewed by Štefan Miklošovič for CASSANDRA-20132

Commit:
Fix failing DatabaseDescriptorRefTest, fix flaky test in ReadCommandTest

Patch by Dmitry Konstantinov; reviewed by Štefan Miklošovič for CASSANDRA-20132
